gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/javaversions/BaselineJavaVersion.java

```diff
@@ -101,127 +101,150 @@ public void execute(CheckJavaVersionsTask task) {
+            JavaToolchains baselineConfiguredJavaToolchains,
+            BaselineJavaVersionsExtension rootExtension,
+            JavaToolchainService javaToolchainService) {
+
+        project.getTasks().withType(JavaCompile.class).configureEach(javaCompileTask -> {
+            setJavaCompiler(
+                    javaCompileTask, rootExtension, baselineConfiguredJavaToolchains, javaToolchainService, target);
+            javaCompileTask.getOptions().getCompilerArgumentProviders().add(new EnablePreviewArgumentProvider(target));
+            // Set sourceCompatibility to opt out of '-release', allowing opens/exports to be used.
+            javaCompileTask.doFirst(task -> ((JavaCompile) task)
+                    .setSourceCompatibility(target.get().javaLanguageVersion().toString()));
         });
 
-        project.getTasks().withType(Javadoc.class).configureEach(new Action<Javadoc>() {
-            @Override
-            public void execute(Javadoc javadocTask) {
-                javadocTask.getJavadocTool().set(javaToolchain.flatMap(BaselineJavaToolchain::javadocTool));
-
-                // javadocTask doesn't allow us to add a CommandLineArgumentProvider, so we do it just in time
-                javadocTask.doFirst(new Action<Task>() {
-                    @Override
-                    public void execute(Task task) {
-                        CoreJavadocOptions options = (CoreJavadocOptions) ((Javadoc) task).getOptions();
-                        if (target.get().enablePreview()) {
-                            // yes, javadoc truly takes a single-dash where everyone else takes a double dash
-                            options.addBooleanOption("-enable-preview", true);
-                            options.setSource(target.get().javaLanguageVersion().toString());
-                        }
-                    }
-                });
-            }
+        project.getTasks().withType(Javadoc.class).configureEach(javadocTask -> {
+            setJavaDocTool(javadocTask, rootExtension, baselineConfiguredJavaToolchains, javaToolchainService, target);
+
+            // javadocTask doesn't allow us to add a CommandLineArgumentProvider, so we do it just in time
+            javadocTask.doFirst(task -> {
+                CoreJavadocOptions options = (CoreJavadocOptions) ((Javadoc) task).getOptions();
+                if (target.get().enablePreview()) {
+                    // yes, javadoc truly takes a single-dash where everyone else takes a double dash
+                    options.addBooleanOption("-enable-preview", true);
+                    options.setSource(target.get().javaLanguageVersion().toString());
+                }
+            });
         });
 
         // checkstyle.getJavaLauncher() was added in Gradle 7.5
         if (GradleVersion.current().compareTo(GradleVersion.version("7.5")) >= 0) {
-            project.getTasks().withType(Checkstyle.class).configureEach(new Action<Checkstyle>() {
```

Comment:
configureEach/stuff that isn't a task action or a CommandLineArgumentsProvider are fine to have as a lambda though